### PR TITLE
function chaining practice done

### PR DESCRIPTION
**Pull request by smjebelli**
Exception handling if the IP is from Islamic Republic!! using chained functions.

### CODE REVIEW

File: src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs [hunk 1]
```diff
@@ -0,0 +1,40 @@
+using Dumpify;
+using PipelineDesignPattern;
+
+public class Framework
+{
+    public class Framewrok()
```
Comment (mohammadKarimi) [inline src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs:6]: useless default constructor.
remove it

File: src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    {
+    }
+    public delegate void Action(HttpContext context);
+    public void ExceptionHandling(HttpContext context, Action function)
+    {
+        try
+        {
+            function(context);
+        }
+        catch (Exception ex)
```
Comment (mohammadKarimi) [inline src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs:16]: this lice captures all type of exception, please develop issue related exceptions

File: src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            "Exception catched: ".Dump(ex.Message);
+        }
+
+    }
+    public void Authentication(HttpContext context, Action function)
+    {
+        context.IP.Dump("Start auth...");
+        if (context.IP.Contains("Iran"))
```
Comment (mohammadKarimi) [inline src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs:25]: IP format is XXX.XXX.XXX.XXX, it does not contain of Iran.

File: src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        {
+            throw new Exception("Sorry you are accessing from Islamic Republic.");
```
Comment (mohammadKarimi) [inline src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs:27]: Throw Issue Related Exception

File: src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        function(context);
+        context.IP.Dump("End auth...");
+    }
+
+    public void Cors(HttpContext context, Action function)
+    {
+        context.IP.Dump("Start Cors...");
+        function(context);
+        context.IP.Dump("End Cors...");
+
```
Comment (mohammadKarimi) [inline src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs:38]: while space, try to write code in pretty format.

File: src/MohammadJebelli/PipelineDesignPattern/HttpContext.cs [hunk 6]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Principal;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace PipelineDesignPattern
+{
+    public class HttpContext
```
Comment (mohammadKarimi) [inline src/MohammadJebelli/PipelineDesignPattern/HttpContext.cs:10]: you can use, record, 
or add required keyword to property.

File: src/MohammadJebelli/PipelineDesignPattern/Program.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+Framework framework = new Framework();
+
+var httpContext1 = new HttpContext() { IP = "Iran" };
+var httpContext2 = new HttpContext() { IP = "US" };
+
+
+"US IP----------------------------".Dump();
+framework.ExceptionHandling(
+    httpContext2, (context) =>
+        framework.Authentication(context, controller.GetAllUser)
```
Comment (mohammadKarimi) [inline src/MohammadJebelli/PipelineDesignPattern/Program.cs:14]: Pretty format for improving readability.

**Review by mohammadKarimi — CHANGES_REQUESTED** [review]

File: src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs (hunk 3 above)
Comment (smjebelli) [reply src/MohammadJebelli/PipelineDesignPattern/Framewrok.cs:25]: the reason I used this approach was to simplify.
Should I use a package like IPInfo to extract IP location?

**Comment by smjebelli** [comment]
Thanks for your review.
Suggested changes applied

**Review by mohammadKarimi — APPROVED** [review]
Perfect

**Pull request merged**